File: core/bin_merger.py
# OpenROM — Universal ROM Compression Suite
# M5 Dev | GPL v3

import os
import re
from typing import Callable, Tuple, List, Dict


def _sectors_to_msf(sectors: int) -> str:
    """Convert a sector count to MSF (MM:SS:FF) format (75 sectors/sec)."""
    mm = sectors // (75 * 60)
    rem = sectors % (75 * 60)
    ss = rem // 75
    ff = rem % 75
    return f"{mm:02d}:{ss:02d}:{ff:02d}"


def _msf_to_sectors(msf: str) -> int:
    """Convert MSF string (MM:SS:FF) to sector count."""
    parts = list(map(int, msf.split(":")))
    return parts[0] * 60 * 75 + parts[1] * 75 + parts[2]


def _get_sector_size(mode: str) -> int:
    """Return sector size in bytes based on track mode."""
    mode_upper = mode.upper()
    if "2048" in mode_upper:
        return 2048
    if "2336" in mode_upper:
        return 2336
    return 2352


def parse_cue(cue_path: str) -> List[Dict]:
    """
    Parse a CUE file to extract tracks, indexes, and referenced BIN files.
    Returns list of track dicts:
      [
        {
          "track_number": int,
          "mode": str,
          "file": str (absolute path),
          "rel_file": str (filename in cue),
          "indexes": [{"number": int, "msf": str}, ...]
        }, ...
      ]
    """
    cue_dir = os.path.dirname(os.path.abspath(cue_path))
    tracks = []

    current_file = None
    file_regex = re.compile(r'FILE\s+["\']?([^"\']+)["\']?\s+BINARY', re.IGNORECASE)
    track_regex = re.compile(r'TRACK\s+(\d+)\s+([^\s]+)', re.IGNORECASE)
    index_regex = re.compile(r'INDEX\s+(\d+)\s+(\d{2}:\d{2}:\d{2})', re.IGNORECASE)

    with open(cue_path, "r", encoding="utf-8", errors="replace") as f:
        for line in f:
            line_str = line.strip()
            fm = file_regex.search(line_str)
            if fm:
                rel_file = fm.group(1)
                current_file = os.path.join(cue_dir, rel_file)
                continue

            tm = track_regex.search(line_str)
            if tm:
                num = int(tm.group(1))
                mode = tm.group(2)
                tracks.append({
                    "track_number": num,
                    "mode": mode,
                    "file": current_file,
                    "rel_file": os.path.basename(current_file) if current_file else "",
                    "indexes": [],
                })
                continue

            im = index_regex.search(line_str)
            if im and tracks:
                idx_num = int(im.group(1))
                idx_msf = im.group(2)
                tracks[-1]["indexes"].append({
                    "number": idx_num,
                    "msf": idx_msf,
                })

    return tracks


def merge_bins(
    cue_path: str,
    output_dir: str = None,
    on_progress: Callable[[float], None] = None
) -> Tuple[str, str]:
    """
    Merge multiple BIN files (from a multi-file BIN/CUE set) into a single BIN + updated CUE.

    Returns:
      (merged_bin_path, new_cue_path)
    """
    if not os.path.exists(cue_path):
        raise FileNotFoundError(f"CUE file not found: {cue_path}")

    cue_dir = os.path.dirname(os.path.abspath(cue_path))
    target_dir = output_dir or cue_dir or "."
    os.makedirs(target_dir, exist_ok=True)

    tracks = parse_cue(cue_path)
    if not tracks:
        raise ValueError(f"No tracks found in CUE file: {cue_path}")

    # Validate all referenced BIN files exist
    for t in tracks:
        bin_file = t.get("file")
        if not bin_file or not os.path.exists(bin_file):
            raise FileNotFoundError(f"Referenced BIN file not found: {bin_file}")

    base_stem = os.path.splitext(os.path.basename(cue_path))[0]
    merged_bin_name = f"{base_stem}_merged.bin"
    merged_cue_name = f"{base_stem}_merged.cue"

    merged_bin_path = os.path.join(target_dir, merged_bin_name)
    merged_cue_path = os.path.join(target_dir, merged_cue_name)

    total_bytes = sum(os.path.getsize(t["file"]) for t in tracks)
    bytes_written = 0

    accumulated_sectors = 0
    new_cue_tracks = []

    # Open merged BIN file for writing
    with open(merged_bin_path, "wb") as out_bin:
        for idx, t in enumerate(tracks):
            bin_path = t["file"]
            mode = t["mode"]
            sec_size = _get_sector_size(mode)

            file_size = os.path.getsize(bin_path)
            sectors_in_file = file_size // sec_size

            parsed_indexes = t.get("indexes", [])

            if parsed_indexes:
                # Find INDEX 01 offset within this track file
                idx01_entry = next((i for i in parsed_indexes if i["number"] == 1), None)
                idx01_offset = _msf_to_sectors(idx01_entry["msf"]) if idx01_entry else 0

                computed_indexes = []
                for idx_item in parsed_indexes:
                    idx_offset = _msf_to_sectors(idx_item["msf"])
                    # Convert track-relative index offset into merged timeline
                    idx_sectors = accumulated_sectors + (idx_offset - idx01_offset)
                    computed_indexes.append({
                        "number": idx_item["number"],
                        "msf": _sectors_to_msf(idx_sectors),
                    })
            else:
                computed_indexes = [{
                    "number": 1,
                    "msf": _sectors_to_msf(accumulated_sectors),
                }]

            new_cue_tracks.append({
                "track_number": t["track_number"],
                "mode": mode,
                "indexes": computed_indexes,
            })

            accumulated_sectors += sectors_in_file

            # Write BIN file contents in chunks
            chunk_size = 1024 * 1024  # 1MB
            with open(bin_path, "rb") as in_bin:
                while True:
                    chunk = in_bin.read(chunk_size)
                    if not chunk:
                        break
                    out_bin.write(chunk)
                    bytes_written += len(chunk)
                    if on_progress and total_bytes > 0:
                        pct = min(100.0, (bytes_written / total_bytes) * 100.0)
                        on_progress(pct)

    if on_progress:
        on_progress(100.0)

    # Write new merged CUE file
    with open(merged_cue_path, "w", encoding="utf-8") as f:
        f.write(f'FILE "{merged_bin_name}" BINARY\n')
        for ct in new_cue_tracks:
            f.write(f'  TRACK {ct["track_number"]:02d} {ct["mode"]}\n')
            for idx_item in ct["indexes"]:
                f.write(f'    INDEX {idx_item["number"]:02d} {idx_item["msf"]}\n')

    return (merged_bin_path, merged_cue_path)

File: core/lib/src/bin_merger.dart
// OpenROM — Universal ROM Compression Suite
// M5 Dev | GPL v3

import 'dart:io';
import 'package:path/path.dart' as p;

String _sectorsToMsf(int sectors) {
  final mm = sectors ~/ (75 * 60);
  final rem = sectors % (75 * 60);
  final ss = rem ~/ 75;
  final ff = rem % 75;
  return '${mm.toString().padLeft(2, '0')}:${ss.toString().padLeft(2, '0')}:${ff.toString().padLeft(2, '0')}';
}

int _msfToSectors(String msf) {
  final parts = msf.split(':').map(int.parse).toList();
  return parts[0] * 60 * 75 + parts[1] * 75 + parts[2];
}

int _getSectorSize(String mode) {
  final modeUpper = mode.toUpperCase();
  if (modeUpper.contains('2048')) return 2048;
  if (modeUpper.contains('2336')) return 2336;
  return 2352;
}

class CueTrackIndex {
  final int number;
  final String msf;

  CueTrackIndex({required this.number, required this.msf});
}

class CueTrack {
  final int trackNumber;
  final String mode;
  final String? file;
  final String relFile;
  final List<CueTrackIndex> indexes;

  CueTrack({
    required this.trackNumber,
    required this.mode,
    required this.file,
    required this.relFile,
    required this.indexes,
  });
}

List<CueTrack> parseCue(String cuePath) {
  final cueDir = p.dirname(p.canonicalize(cuePath));
  final tracks = <CueTrack>[];

  String? currentFile;
  final fileRegex = RegExp('FILE\\s+["\']?([^"\']+)["\']?\\s+BINARY', caseSensitive: false);
  final trackRegex = RegExp('TRACK\\s+(\\d+)\\s+([^\\s]+)', caseSensitive: false);
  final indexRegex = RegExp('INDEX\\s+(\\d+)\\s+(\\d{2}:\\d{2}:\\d{2})', caseSensitive: false);

  final lines = File(cuePath).readAsLinesSync();
  for (final line in lines) {
    final lineStr = line.trim();
    final fm = fileRegex.firstMatch(lineStr);
    if (fm != null) {
      final relFile = fm.group(1)!;
      currentFile = p.join(cueDir, relFile);
      continue;
    }

    final tm = trackRegex.firstMatch(lineStr);
    if (tm != null) {
      final num = int.parse(tm.group(1)!);
      final mode = tm.group(2)!;
      tracks.add(CueTrack(
        trackNumber: num,
        mode: mode,
        file: currentFile,
        relFile: currentFile != null ? p.basename(currentFile) : '',
        indexes: [],
      ));
      continue;
    }

    final im = indexRegex.firstMatch(lineStr);
    if (im != null && tracks.isNotEmpty) {
      final idxNum = int.parse(im.group(1)!);
      final idxMsf = im.group(2)!;
      tracks.last.indexes.add(CueTrackIndex(number: idxNum, msf: idxMsf));
    }
  }

  return tracks;
}

class BinMergeResult {
  final String mergedBinPath;
  final String mergedCuePath;

  BinMergeResult(this.mergedBinPath, this.mergedCuePath);
}

BinMergeResult mergeBins(
  String cuePath, {
  String? outputDir,
  void Function(double pct)? onProgress,
}) {
  if (!File(cuePath).existsSync()) {
    throw FileSystemException('CUE file not found: $cuePath', cuePath);
  }

  final cueDir = p.dirname(p.canonicalize(cuePath));
  final targetDir = (outputDir != null && outputDir.isNotEmpty) ? outputDir : cueDir;
  Directory(targetDir).createSync(recursive: true);

  final tracks = parseCue(cuePath);
  if (tracks.isEmpty) {
    throw FormatException('No tracks found in CUE file: $cuePath');
  }

  for (final t in tracks) {
    final binFile = t.file;
    if (binFile == null || !File(binFile).existsSync()) {
      throw FileSystemException('Referenced BIN file not found: $binFile', binFile);
    }
  }

  final cueName = p.basename(cuePath);
  final idx = cueName.lastIndexOf('.');
  final baseStem = idx != -1 ? cueName.substring(0, idx) : cueName;

  final mergedBinName = '${baseStem}_merged.bin';
  final mergedCueName = '${baseStem}_merged.cue';

  final mergedBinPath = p.join(targetDir, mergedBinName);
  final mergedCuePath = p.join(targetDir, mergedCueName);

  int totalBytes = 0;
  for (final t in tracks) {
    totalBytes += File(t.file!).lengthSync();
  }
  int bytesWritten = 0;

  int accumulatedSectors = 0;
  final newCueTracks = <Map<String, dynamic>>[];

  final outBinFile = File(mergedBinPath);
  final outSink = outBinFile.openSync(mode: FileMode.write);

  try {
    for (final t in tracks) {
      final binPath = t.file!;
      final mode = t.mode;
      final secSize = _getSectorSize(mode);

      final fileSize = File(binPath).lengthSync();
      final sectorsInFile = fileSize ~/ secSize;

      final parsedIndexes = t.indexes;

      List<CueTrackIndex> computedIndexes;
      if (parsedIndexes.isNotEmpty) {
        CueTrackIndex? idx01Entry;
        for (final item in parsedIndexes) {
          if (item.number == 1) {
            idx01Entry = item;
            break;
          }
        }
        final idx01Offset = idx01Entry != null ? _msfToSectors(idx01Entry.msf) : 0;

        computedIndexes = [];
        for (final idxItem in parsedIndexes) {
          final idxOffset = _msfToSectors(idxItem.msf);
          final idxSectors = accumulatedSectors + (idxOffset - idx01Offset);
          computedIndexes.add(CueTrackIndex(
            number: idxItem.number,
            msf: _sectorsToMsf(idxSectors),
          ));
        }
      } else {
        computedIndexes = [
          CueTrackIndex(number: 1, msf: _sectorsToMsf(accumulatedSectors)),
        ];
      }

      newCueTracks.add({
        'trackNumber': t.trackNumber,
        'mode': mode,
        'indexes': computedIndexes,
      });

      accumulatedSectors += sectorsInFile;

      final inRaf = File(binPath).openSync(mode: FileMode.read);
      const chunkSize = 1024 * 1024;
      while (true) {
        final chunk = inRaf.readSync(chunkSize);
        if (chunk.isEmpty) break;
        outSink.writeFromSync(chunk);
        bytesWritten += chunk.length;
        if (onProgress != null && totalBytes > 0) {
          final pct = (bytesWritten / totalBytes * 100.0).clamp(0.0, 100.0);
          onProgress(pct);
        }
      }
      inRaf.closeSync();
    }
  } finally {
    outSink.closeSync();
  }

  if (onProgress != null) {
    onProgress(100.0);
  }

  final cueBuffer = StringBuffer();
  cueBuffer.writeln('FILE "$mergedBinName" BINARY');
  for (final ct in newCueTracks) {
    final trackNum = (ct['trackNumber'] as int).toString().padLeft(2, '0');
    final mode = ct['mode'] as String;
    cueBuffer.writeln('  TRACK $trackNum $mode');
    final indexes = ct['indexes'] as List<CueTrackIndex>;
    for (final idxItem in indexes) {
      final idxNum = idxItem.number.toString().padLeft(2, '0');
      cueBuffer.writeln('    INDEX $idxNum ${idxItem.msf}');
    }
  }

  File(mergedCuePath).writeAsStringSync(cueBuffer.toString());

  return BinMergeResult(mergedBinPath, mergedCuePath);
}
